### PR TITLE
Num index pd df

### PR DESCRIPTION
**Pull request by HenryLeongStat**
Ready to merge

### CODE REVIEW

File: src/sos_r/kernel.py [hunk 1]
```diff
@@ -63,7 +63,17 @@ def _R_repr(obj):
     elif obj is None:
         return 'NULL'
     elif isinstance(obj, dict):
-        return 'list(' + ','.join('{}={}'.format(x, _R_repr(y)) for x,y in obj.items()) + ')'
+        # check whether each key string is a valid string in R
+        import re
+        _sos_dealing_with_invalid_key_pattern = re.compile("\W|^(?=\d)")
```
Comment (HenryLeongStat) [inline src/sos_r/kernel.py]: Probably it is a bad function name?

**Comment by HenryLeongStat** [comment]
Professor, I still can't see your comments on the code. But I made the changes based on your screen shot comments.

**Comment by BoPeng** [comment]
>  def make_name(input_string):

Move this outside because you do not want to define this function many many times.

> re.compile("\W|^(?=\d)")

The second part is not needed because it detects if the string is lead by character and we have done that.


> +invalid_key_pattern = re.compile("\W|^(?=\d)")

pre-compilation is no longer needed if we are dealing with a very simple expression now. (`\W`).

**Comment by HenryLeongStat** [comment]
Understood!

**Comment by BoPeng** [comment]
You called `str` twice.

The comment and the last sub logic of the function are wrong.

**Comment by HenryLeongStat** [comment]
>The comment

How about `# check whether each key string is a valid string in R`  ->  `# make the SoS dict key name to be valid in R list`?

> and the last sub logic of the function are wrong

Do you mean `return re.sub('\W', '_', name)`? We should still take care of the special characters (such as '(', ')' ), isn't it?

**Comment by BoPeng** [comment]
> Do you mean return re.sub('\W', '_', name)? 

indentation. Now you are only replacing invalid character for name with leading number.

**Comment by HenryLeongStat** [comment]
Oops... My bad... Will fix it now!

**Comment by BoPeng** [comment]
> name = input_string

What is the point of this?

Also, the tests still fail.

**Comment by HenryLeongStat** [comment]
>name = input_string

My bad... Fixed the `str()` and didn't see this line... 😅 

The test is weird... It passes on my machine:

```
$ python test_R_kernel.py
/Users/manchongleong/anaconda/lib/python3.6/site-packages/jupyter_client/manager.py:72: DeprecationWarning: KernelManager._kernel_name_changed is deprecated in traitlets 4.1: use @observe and @unobserve instead.
  def _kernel_name_changed(self, name, old, new):
./Users/manchongleong/anaconda/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
....
----------------------------------------------------------------------
Ran 5 tests in 17.341s

OK
```

Including the test I added, there are 5 tests in total. But how come it fails in Travis CI build?

**Comment by BoPeng** [comment]
You forgot not handle a corner case `{'': 1}`.

**Comment by HenryLeongStat** [comment]
😅 Got it fixed!
But Professor, one potential issue here is that, what if an extreme case {'' : 1, 'X' : 2} happens? Should I take care of this case? I mean, 'X' as key should be a common case.

**Comment by HenryLeongStat** [comment]
OK... In this case, the dimensions of a R list with same 'key' will lose meeting for 'key searching' because the second (or third or ...) couldn't be visited with {listname}${dimname} or {listname}[r'{dimname}']...
